server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/priority/SmtpMtPriorityParameterHook.java

```diff
@@ -0,0 +1,77 @@
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.smtpserver.priority;
+
+import static org.apache.james.protocols.api.ProtocolSession.State.Transaction;
+import static org.apache.james.smtpserver.priority.SmtpMtPriorityParameters.MT_PRIORITY_PARAMETER;
+
+import java.util.Optional;
+
+import org.apache.james.protocols.api.ProtocolSession;
+import org.apache.james.protocols.smtp.SMTPRetCode;
+import org.apache.james.protocols.smtp.SMTPSession;
+import org.apache.james.protocols.smtp.hook.HookResult;
+import org.apache.james.protocols.smtp.hook.HookReturnCode;
+import org.apache.james.protocols.smtp.hook.MailParametersHook;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SmtpMtPriorityParameterHook implements MailParametersHook {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SmtpMtPriorityParameterHook.class);
+
+    public static final ProtocolSession.AttachmentKey<SmtpMtPriorityParameters.MtPriority> MT_PRIORITY_ATTACHMENT_KEY =
+        ProtocolSession.AttachmentKey.of(MT_PRIORITY_PARAMETER, SmtpMtPriorityParameters.MtPriority.class);
+
+    @Override
+    public HookResult doMailParameter(SMTPSession session, String paramName, String paramValue) {
+        if (session.getAttachment(MT_PRIORITY_ATTACHMENT_KEY, Transaction).isPresent()) {
+            LOGGER.debug("The Mail parameter cannot contain more than one MT-PRIORITY parameter at the same time");
+            return HookResult.builder()
+                .smtpReturnCode(SMTPRetCode.SYNTAX_ERROR_ARGUMENTS)
+                .hookReturnCode(HookReturnCode.deny())
+                .smtpDescription("The Mail parameter cannot contain more than one MT-PRIORITY parameter at the same time")
+                .build();
+        }
+        if (paramName.equals(MT_PRIORITY_PARAMETER)) {
+            try {
+                Optional<SmtpMtPriorityParameters.MtPriority> mtPriority = SmtpMtPriorityParameters.MtPriority.from(paramValue);
+                if (mtPriority.filter(p -> session.getUsername() == null && p.getPriorityValue() > 0).isPresent()) {
```

Comment:
Who is able to set which priory is actually quite crucial and saying "high priorities are only for connected users" seems somehow arbitrary. (We can be more severe than the spec and allow eg some high priority just to some individuals)

I would encapsulate the check into a dedicated method of smtp session - so that it could get easily overloaded if need be:

```
default boolean isAuthorized(MtPriority p) { 
     return session.getUsername() != null && p.getPriorityValue() <= 0;
}
```

And then:

```
mtPriority.ifPresent(p -> {
    if (!session.isAuthorized(p)) {
        // reject
    }
    session.setAttachment(MT_PRIORITY_ATTACHMENT_KEY, p, Transaction))
});
```

(possibly even extracting the resulting lambda as a method)